vernac/prettyp.ml

```diff
@@ -929,6 +929,32 @@ let print_name env sigma na udecl =
   | {loc; v=Constrexpr.AN ref} ->
     print_any_name env sigma (locate_any_name ref) udecl
 
+let print_notation_grammar env sigma ntn =
+  let ng = List.hd (Notgram_ops.grammar_of_notation ntn) in
+  let assoc = ng.Notation_gram.notgram_assoc in
+  let prdf () = Pp.str "No associativity defined" in
+  Pp.(pr_opt_default prdf Gramlib.Gramext.pr_assoc assoc)
+
+exception PrintNotationNotFound of string
+
+let () = CErrors.register_handler (function
+  | PrintNotationNotFound ntn_str ->
+      Some Pp.(str "Notation" ++ spc () ++ str "\""
+        ++ str ntn_str ++ str "\"" ++ spc () ++ str "was not found." ++ spc ()
+        ++ str "Make sure all the holes are explicitly indicated with _.")
+  | _ -> None)
+
+let print_notation env sigma raw_ntn =
+  (* convert notation string to key. eg. "x + y" to "_ + _" *)
+  let ntn = (Constrexpr.InConstrEntry, Notation.interpret_notation_string raw_ntn) in
+  try
+    let lvl = Notation.level_of_notation ntn in
+    let args = Notgram_ops.subentries_of_notation ntn in
+    let pplvl = Metasyntax.pr_level ntn lvl args in
+    Pp.(str "Notation " ++ pplvl ++ pr_comma ()
+      ++ print_notation_grammar env sigma ntn ++ spc () ++ str "associativity")
+  with Not_found -> raise @@ PrintNotationNotFound raw_ntn
```

Comment:
Where's the not_found raised? Maybe we could delimit the scope a bit more. [Or even better introduce a _opt function so types can know about the not found]

Reply:
I need to check, but I suspect it can happen in `Notation.level_of_notation` and probably `Notgram_ops.subentries_of_notation` too. Here the `ntn` key is queried in a dictionary which usually raises the `Not_found`.